### ISSUE TITLE
Implement Web (HTTP) Seeding (BEP17+BEP19)

### ISSUE THREAD

**Issue by fisch0920** [feross/webtorrent#67]
Web / HTTP seeding is an extension to standard bittorrent p2p which allows a client to download ranges from normal HTTP servers in addition to standard bittorrent peers.  This has the advantage of effectively adding permanent, seed-only peers to any swarm, and it's very easy for any existing HTTP fileservers to facilitate bittorrent and also reduce their own load by offloading portions of normal HTTP downloads to bittorrent peers.

I think this feature will be **particularly important to the browser version**, since we're expecting the number of hybrid peers to be relatively low compared with normal bittorrent clients, and HTTP-based peers will be supported in the browser without any workarounds.

Note that there are two competing standards for HTTP seeding, both with their own pros / cons, and I believe most clients implement them both since neither is very complex.  The nomenclature used to refer to the standards is very inconsistent, so for the purposes of webtorrent, I propose we refer them simply as [BEP19](http://bittorrent.org/beps/bep_0019.html) (aka GetRight-style http/ftp seeding) and [BEP17](http://www.bittorrent.org/beps/bep_0017.html) (aka Hoffman-style, http seeding, webseeding).

BEP19 has the advantage of being able to work with *any* file accessible via normal HTTP or FTP, utilizing standard HTTP range requests to download pieces (or better yet reserved spans of pieces to minimize separate HTTP requests).  BEP17 requires external servers to be modified to support BEP17, which is a disadvantage in terms of interoperability, but it does have the advantage of allowing the seeding servers to better control traffic and avoid bandwidth abuse.

This task will require modifying `parse-torrent` to recognize HTTP seed servers.  It will also require changes to `bittorrent-swarm`, possibly `bittorrent-protocol`, and `bittorrent-client` if we want to support more efficient utilization of HTTP by reserving spans of pieces for each individual HTTP seed request as opposed to treating HTTP seed peers as normal peers and requesting single blocks at a time from them.

Resources:
* High level [overview](http://en.wikipedia.org/wiki/BitTorrent#Web_seeding) of HTTP seeding and protocols
* Vuze wiki's [overview](http://wiki.vuze.com/w/HTTP_Seeding) of its HTTP seeding implementation

**Comment by olalonde** [comment]
I started writing an implementation (https://github.com/feross/webtorrent/issues/331, https://github.com/feross/bittorrent-swarm/pull/19). Basic functionality implemented.

**Comment by feross** [comment]
Thanks to @olalonde's PR, we have support for BEP19 in in node!

The approach isn't very efficient since it makes separate requests for each 16KB block, but that's okay for now. We have a starting point to work from! This is released as webtorrent 0.51.0.

**Comment by feross** [comment]
Multi-file web seed doesn't work.

@olalonde The sponsor confirmed they can still donate the $500.

**Comment by olalonde** [comment]
@feross great. I've been busy with other things lately but will take a stab at it as soon as I free up some time.

**Comment by solderzzc** [comment]
Adding the multiple files support on this pull request https://github.com/feross/bittorrent-swarm/pull/24

**Comment by feross** [comment]
@solderzzc Thanks - please see my comments on the PR.

**Comment by bertofer** [comment]
Any news on that? I would love to use this feature, i'm working on a project to distribute files already uploaded to the server via webtorrent, and this would be very nice.

**Comment by jakefb** [comment]
@bertofer you can put &ws= it in the magnet link like this:

`var magnetUri = 'magnet:?xt=urn:btih:ab3f1350ebe4563a710545d0e33e09a7b7943ecf&tr=wss://tracker.webtorrent.io&ws=http%3A%2F%2Ffastcast.nz%2Fdownloads%2Fawakening-new-zealand-4k.mp4'`

The URLs to the files uploaded on your server has to be encoded. You can use http://meyerweb.com/eric/tools/dencoder/ to do this.

As far as I understand, if you use a magnet link WebTorrent will start downloading the files from the web seed once a WebRTC peer (https://instant.io) has connected and sent the torrent metadata. You can then stop seeding from the WebRTC peer and the web seed will send over the rest of the file. You can see this working in my project http://fastcast.nz if you like :)

Are you trying to get this to work in the browser? I would be interested in checking out your project! I'm trying to work out how to use a .torrent file in the browser instead of a magnet link, so the torrent instantly starts downloading from the web seed https://github.com/feross/webtorrent/issues/424.

**Comment by bertofer** [comment]
@jakefb Thanks for your answer!! 

Yes, i want it in the browser. So, to download a folder, via webseed, you must use a url that sends you to those ugly autoindex of files in a folder via http? Imagine your downloads forlder is a torrent, can i use that url? http%3A%2F%2Ffastcast.nz%2Fdownloads%2? 

My project is filelibrary, check it out on my profile :) for now i'm only testing the platform so expect ugly code and ugly everything :P What i want to achieve is to distribute files via browser with torrenting, but also an interface to upload that files to the server, and being the server a first seed for that file, to ensure always a seed active (For that i wanted webseed :D ). The torrents are generated once the file uploaded, and saved on MongoDB (As buffer), so the webpage that downloads it can also get the .torrent to start downloading (as a Buffer).

**Comment by solderzzc** [comment]
Lol my multiple files web seeds patch is still waiting for merging. The
test failed since the patch in
https://github.com/feross/bittorrent-swarm/pull/24 is not merged.

**Comment by jakefb** [comment]
@bertofer I have managed to parse a torrent file in the browser so it starts downloading from the web seed instantly! Have a look at the code here https://github.com/fastcast/fastcast/blob/master/client/stream-torrent.js. You can see it in action on one of my videos such as http://fastcast.nz/videos/awakening-new-zealand-4k.html. The video also loads instantly when using a torrent file instead of a magnet link, which is very convenient! This is because it doesn't have to wait for the torrent metadata to be downloaded from the WebRTC peers. I think @feross will find this cool!

To answer your question about having a webseed for a torrent with multiple files, you'll have to wait for @solderzzc PR to be merged https://github.com/feross/webtorrent/pull/397. I have been wondering why it was taking so long for it to get merged, hopefully this happens soon...

**Comment by bertofer** [comment]
Hei @jakefb Thanks for confirming it can be done and works :) Will try it these days! I'll move forward with the project supporting only one file for the moment.

**Comment by bertofer** [comment]
I'm trying to use a node http-server module to serve the file as a webseed, but i got this when downloading from webtorrent:

bittorrent-swarm:peer destroy http://localhost:9000/TinyCore-current.iso (error: handshake timeout) +25s
browser.js:113 bittorrent-protocol [94934491] destroy +0ms
browser.js:113 bittorrent-protocol [94934491] end +1ms
browser.js:113 bittorrent-protocol [94934491] got uninterested +1ms
browser.js:113 bittorrent-swarm:webconn uninterested +0ms
browser.js:113 bittorrent-protocol [94934491] got choke +1ms
browser.js:113 bittorrent-swarm:webconn choke +0ms
browser.js:113 bittorrent-protocol [5af2893b] destroy +0ms

With the debug. Any tip on how to configure the http-server to use it as a web seed? I'm searching and i'm stuck finding any config or requirement to work as a webseed.

Oh, and i'm getting this error also in the middle of the debug

Uncaught TypeError: Cannot read property 'length' of undefined
Wire.handshake @ index.js:191(anonymous function) @ webconn.js:27EventEmitter.emit @ events.js:89Wire._onHandshake @ index.js:392(anonymous function) @ index.js:666Wire._write @ index.js:556doWrite @ writable.js:239writeOrBuffer @ writable.js:229Writable.write @ writable.js:200write @ readable.js:578flow @ readable.js:587(anonymous function) @ readable.js:555(anonymous function) @ browser.js:25

Any ideas?

**Comment by olalonde** [comment]
Does your server support byte range requests? They are required for web seeding.

**Comment by bertofer** [comment]
Yes. I post an image from the "Network" tab of chrome. It returns the 206 Code, and in the headers i can see which byte range was downloaded

![Screeshot](http://s18.postimg.org/4nkf7kkc9/Captura_de_pantalla_2015_09_14_a_las_8_03_01.png)

**Comment by brusacco** [comment]
Get stuck in a CORS problem, the assets I need to "web seed" are in another domain, do I have to add access headers to all sites I want to use as a "web seeder". Thanks works great!

**Comment by jakefb** [comment]
@brusacco Yes, you need to enable CORS on your webserver. [Fastcast](https://github.com/fastcast/fastcast) runs nginx and supports CORS, so you can try downloading http://fastcast.nz/torrents/awakening-new-zealand-4k.torrent for example in your browser  with a xhr request and it will work. You can also fork the Fastcast repository and test this when run from [http-server](https://www.npmjs.com/package/http-server) :D

**Comment by brusacco** [comment]
Thanks it works very well, tested on my own webservers from Chrome. Reading the logs you can see the HTTP range requests .... the Chrome console too.

**Comment by brusacco** [comment]
@jakefb do you know a way to catch that CORS error, and then fallback to a normal html5 video tag from URL, no torrent one. And by the way ... do you note a very high CPU and memory usage by Chrome? In some test I have done, Chrome just crashed after a while.

**Comment by DiegoRBaquero** [comment]
@jakefb Hey! I was looking through webseeding and everything you’ve contributed through fastcast has been a ton of help. I was wondering how you got the webseed urls in the .torrent files. I’m missing that part to remove the need of a webrtc peer. Could you tell how me how you did it? Thank you

**Comment by olalonde** [comment]
@DiegoRBaquero yes, the web seeds need to be listed in the .torrent file. See https://github.com/feross/create-torrent#api (urlList property).

**Comment by olalonde** [comment]
@feross I was thinking of using https://github.com/facebook/dataloader's approach for combining multiple HTTP byte requests into a single one. That is, record all calls to `WebConn.prototype.httpRequest` that happen in an event loop "tick" and on next tick, combine adjacent byte requests into larger ones. For example, webtorrent [seems to call](https://github.com/feross/webtorrent/blob/master/lib/torrent.js#L891)  `wire.request(index, chunkOffset, chunkLength, cb)` synchronously in a while loop. WebConn could capture those requests, e.g.:

```
wire.request(index, 0, 10, cb)
wire.request(index, 10, 10, cb)
wire.request(index, 20, 10, cb)
wire.request(index, 30, 10, cb)
```

and on next tick, combine those into a single

```
self.httpRequest(index, 0, 40, callback)
```

This wouldn't be visible from webtorrent of course but only internal to how webconn works:

```
function WebConn (url, parsedTorrent) {
//...
  this.batch = { requests: [], pending: false };

  self.on('request', function (pieceIndex, offset, length, callback) {
    self.batch.requests.push([ pieceIndex, offset, length, callback ]);
    if (!self.batch.pending) {
      self.batch.pending = true;
      process.nextTick(function () {
        self.batch.pending = false;
        self.processBatch();
      });
    }
  })
//...
}

WebConn.prototype.processBatch = function () {
   // merge this.batch.requests into bigger ones...
   // call this.httpRequest with those bigger requests
   // split "big" responses and pass to original callbacks
}
```

Does that make sense?

**Comment by FluorescentHallucinogen** [comment]
Is FTP and FTPS supported for "&as=" and "&ws=" parameters in magnet?

**Comment by DiegoRBaquero** [comment]
@FluorescentHallucinogen No, a web seed must support HTTP GET requests with the range parameter. Also, from the browser, the server must have CORS enabled.

**Comment by FluorescentHallucinogen** [comment]
OK, FTP(S) support is not implemented in WebTorrent at this moment. But is it possible to implement FTP(S) support in the future?

Most common web browsers can retrieve files hosted on FTP servers. In other words, web browsers already have FTP support implemented. :)

Not sure about FTPS support in web browsers... Is Google Chrome or Mozilla Firefox supports FTPS?

Many FTP clients and download managers support resume downloading and multipart downloading over FTP (if the server support these features).

For HTTP you can use byte-ranges to resume anywhere or download specific ranges you specify, but with FTP you can only say where to start the download.

Not sure about support of resume downloading and multipart downloading over FTP in web browsers... Is Google Chrome and Mozilla Firefox support resume downloading and multipart downloading over FTP?

Anyway, FTP servers can act as a good permanently available "streaming" (sequential mode) seeds.

What about CORS for FTP?

**Comment by fisch0920** [comment]
FTP doesn't have anything to do with bittorrent and is not supported by any bittorrent engine I'm aware of.  The only reason web seeding is supported at all is because of the BEPs linked to in my original post (bittorrent specs).

Everyone, let's please keep this thread specifically about webseeding via BEP17+BEP19.

@olalonde I know you posted your gist awhile ago, but I do like its simplicity.

**Comment by FluorescentHallucinogen** [comment]
FTP seeding is described in BEP19.

GetRight supports FTP seeding.

The browser can download at any point (or at the beginning at worst case) of the file via FTP and seed (upload) via BitTorrent. Are there any technical limitations to not implement FTP support in WebTorrent?

**Comment by fisch0920** [comment]
My bad; it's been awhile since I've looked at them :)

**Comment by olalonde** [comment]
I don't think it's possible to do FTP in the browser so that would be a Node.js only feature if it gets implemented.

**Comment by rom1504** [comment]
it's quite possible. try this ftp://speedtest.tele2.net/

**Comment by olalonde** [comment]
AFAIK some browsers double as an FTP client but there is no standard way to download FTP files from Javascript.

**Comment by DiegoRBaquero** [comment]
@FluorescentHallucinogen In node it would be possible. It would just need to be implemented. But I don't see why not use HTTP instead. 

In the browser it is not possible, FTP doesn't have cors and XHR (AJAX) doesn't support it. See [this](http://stackoverflow.com/questions/14839838/what-is-the-syntax-to-do-a-cross-domain-xmlhttprequest-to-an-ftp-server)

**Comment by FluorescentHallucinogen** [comment]
Are there any diffences between `&as=` and `&ws=` parameters in magnet links? Is `&as=` an alias of `&ws=` **in every respects**?

I know that `&ws=` parameter according to the specification supports links to directories (folders) for multi-file torrents.

Does `&as=` parameter **according to its specification** supports links to directories (folders) or only to single files?

**Comment by FluorescentHallucinogen** [comment]
If one of the trackers in the magnet URI e.g. `magnet:?xt=urn:btih:df2a06c6d602f5514f11246dcb843efa403efe0c&dn=amplitude-new-zealand-4k.mp4&tr=ws%3A%2F%2Ftracker.fastcast.nz&tr=wss%3A%2F%2Ftracker.btorrent.xyz&tr=wss%3A%2F%2Ftracker.openwebtorrent.com&tr=wss%3A%2F%2Ftracker.webtorrent.io&ws=http%3A%2F%2Ffastcast.nz%2Fdownloads%2Famplitude-new-zealand-4k.mp4&ws=https%3A%2F%2Fwebseed.btorrent.xyz%2Famplitude-new-zealand-4k.mp4` is insecure (starting 'ws://') e.g `ws://tracker.fastcast.nz` the **whole** process of downloading torrent from the page loaded over HTTPS e.g. https://instant.io or https://btorrent.xyz is blocked by the browser. Connecting to the rest of secure trackers from the magnet URI (starting `wss://`) e.g. `wss://tracker.btorrent.xyz`, `wss://tracker.openwebtorrent.com`, `wss://tracker.webtorrent.io` does not occur.

Is it possible to implement in WebTorrent avoiding connecting to a non-secure WebSocket server from a secure origin?

**Comment by sirajea** [comment]
@FluorescentHallucinogen There is very little documentation. I found https://wiki.theory.org/BitTorrent_Magnet-URI_Webseeding and https://en.wikipedia.org/wiki/Magnet_URI_scheme#Web_links_to_the_file 
AS stands for acceptable source (fall-back source), XS for exact source. WS stands for web source/ web seed? The parameters can also be used to retrieve metadata (the .torrent file) instead of the actual payload. Webtorrent's magnet-uri doesn't seem to be using the exact source https://github.com/feross/magnet-uri/blob/master/index.js#L79-L85 and WS overrides AS.

**Comment by feross** [comment]
@FluorescentHallucinogen the issue you described with insecure `ws://` trackers crashing the client on secure `https://` origins is now fixed.

**Comment by FluorescentHallucinogen** [comment]
> AS stands for acceptable source (fall-back source), XS for exact source. WS stands for web source/ web seed? The parameters can also be used to retrieve metadata (the .torrent file) instead of the actual payload.

Does it mean that it is possible to use `&as=` and/or `&xs=` and/or `&ws=` as a link to *.torrent file in the magnet URIs?

For example:

`magnet:?xt=urn:btih:6a9759bffd5c0af65319979fb7832189f4f3c35d
&as=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2Fsintel.torrent`

and/or

`magnet:?xt=urn:btih:6a9759bffd5c0af65319979fb7832189f4f3c35d
&xs=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2Fsintel.torrent`

and/or

`magnet:?xt=urn:btih:6a9759bffd5c0af65319979fb7832189f4f3c35d
&ws=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2Fsintel.torrent`

Does WebTorrent support using `&as=` and/or `&xs=` and/or `&ws=` as a link to *.torrent file?

**Comment by feross** [comment]
@FluorescentHallucinogen You should use `&ws=`.

**Comment by feross** [comment]
Closing this now, as multi-file web seeds were the last thing we needed to support web seeding fully in WebTorrent. Note that we chose to support [BEP19](http://bittorrent.org/beps/bep_0019.html) and not BEP17, since the former is more web-friendly. I haven't seen any torrents that use the BEP17 style.

**Comment by FluorescentHallucinogen** [comment]
> You should use `&ws=`.

@feross Do you mean `&ws=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2Fsintel.torrent`?

**Comment by feross** [comment]
Sorry for the confusion. You should use `&ws=` to specify the http URL where the files in the torrent are. See this spec: http://www.bittorrent.org/beps/bep_0019.html

We don't support specifying the torrent file location in magnet links at the moment, but I recently discovered this spec (https://wiki.theory.org/BitTorrent_Magnet-URI_Webseeding) that explains how to interpret all these parameters correctly.

There's a new issue here that will track this feature: https://github.com/feross/webtorrent/issues/714

**Comment by JesseWeinstein** [comment]
I'm looking into supporting WebTorrent at the Internet Archive, and ran into a (minor) missing feature in the support for this. Specifically, it appears that WebTorrent doesn't respect the "name" field in .torrent files, which means it looks in the wrong place when trying to load multi-file webseeds that define the "name" field. 

https://wiki.theory.org/BitTorrentSpecification#Info_in_Multiple_File_Mode -- definition of the "name" field
https://archive.org/download/mbid-3124c085-4665-429b-a188-bef1d8aa0a25/mbid-3124c085-4665-429b-a188-bef1d8aa0a25_archive.torrent -- example of a torrent that uses it (and webseeds).

**Comment by feross** [comment]
@JesseWeinstein It's awesome that you're investigating supporting WebTorrent at The Internet Archive! That's been one of my primary goals in developing WebTorrent from the very beginning! Do let me know if there's any way that I can help.

So, about the web seed issue you saw. @dcposch added support for web seeds about a month ago, but it looks like instant.io hasn't been redeployed since then. I just pushed the latest code to it. Let me know if it's working for you now.

You might need to run a local copy of Instant.io since the production site uses Strict Transport Security to force HTTPS and that causes HTTP web seeds to fail with Mixed Content. This is a use case that the standards bodies have not considered. A secure HTTPS application that knows what content it is expecting should be allowed to request data over HTTP.

**Comment by KrishnaPG** [comment]
Looking at web-seeding - is the https://webtorrent.io using it?

The reason I am asking is, on the site's homepage there is a sintel torrent file being downloaded. Now, to test it, I have opened the same site in multiple tabs on the browser.

What I was expecting is, that the new tabs will automatically pickup the previous open tabs as (local) seeders for the content - but that did not happen. Every new tab (page instance) tried to hookup for its own peers from outside, essentially downloading multiple copies of the movie from scratch from same peers.

Theoritically,

1.  each loaded page instance should become a seeder capable of serving other local page load requests (from other tabs or browser instances)
2.  the local page instances (being on the same machine) should become some kind of "local swarm" that partition the remaining work into different chunks and each dowload exclusive portions separately from the "global swarm" and once done spread them equally locally. 

For example, as of now 3 tab instances are downloading bytes 0 to 120MB all 3 times separately (without any knowledge of local peers doing the same thing).

What I would love to see is: 3 tabs each divide the remaining job into 3 pieces (0 to 60MB, 61to 90MB, 91 to 120MB etc.) and each connect to entirely different set of peers and finally merge the content locally. If a new tab opens it should reuse the old content from others are they all re-divide their work (based on remaining portion and completion status)

Is this kind of thing possible with WebTorrent as of now?

**Comment by DiegoRBaquero** [comment]
@KrishnaPG 
> Looking at web-seeding - is the https://webtorrent.io using it?

Yes, it is, but everything is downloaded to memory, no persistent chunk store is being used at this moment.

> What I was expecting is, that the new tabs will automatically pickup the previous open tabs as (local) seeders for the content

That would depend more on the tracker, but it'd would definitely be nice. I've opened an issue [here](https://github.com/feross/bittorrent-tracker/issues/165).

> What I would love to see is: 3 tabs each divide the remaining job into 3 pieces (0 to 60MB, 61to 90MB, 91 to 120MB etc.) and each connect to entirely different set of peers and finally merge the content locally. If a new tab opens it should reuse the old content from others are they all re-divide their work (based on remaining portion and completion status)

That would require syncing information between the 3 tabs. Although it can be done, it's not on the scope of WebTorrent itself. What should be done is connecting the 3 tabs between each other so they can share the pieces they already have at local-link speed. (This already happens when you are the only peer)

**Comment by feross** [comment]
@KrishnaPG What @DiegoRBaquero said.

Also, web seeds are supported fully now, so we can finally close this issue.

**Comment by KrishnaPG** [comment]
Thank you. Just to double confirm, each web-browser instance acts as a potential seed for any new near-by clients, correct?

Which means, if 10 people from near-by location connect to one web-site, any new further connections from near-by locations have the chance of picking up the content from one of these previous browser instances (rather than from the server directly) - correct? Kind of browser-based peer-to-peer CDN

**Comment by feross** [comment]
@KrishnaPG Yes, torrents + web seeds are a form of peer-assisted CDN.

But physical proximity (i.e. nearby peers) has nothing to do with how peers discover each other. The tracker server gives out peers randomly, and the client keeps peers that are fast and cooperating. The http server (web seed) is treated as a peer too, albeit a really good one.
